cuDF/libcudf exponentially weighted moving averages

brandon-b-miller opened the pull request:
Adds an exponentially weighted moving average aggregation to `cudf::scan` and plumbs it up through `cudf.Series.ewm`, similar to `pandas.Series.ewm`.

partially resolves https://github.com/rapidsai/cudf/issues/1263

brandon-b-miller commented:
As I hone in on an implementation that makes sense here I wanted to get some thoughts on pandas nan handling and to what extent / in what layer that null handling should be replicated. Firstly, there is only really `nan` handling in pandas, the new nullable datatypes don't seem to support it yet:

```
ValueError: cannot convert to 'float64'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype
```

The `nan` handling is a little tricky for two reasons. First the simpler thing to deal with: in the result, `nan`s are filled from the last valid value forward. So for the following values you'd have the following result for the moving average

```
>>> pd.Series([1, float('nan'), 3]).ewm(alpha=0.9).mean()
0    1.000000
1    1.000000
2    2.980198
dtype: float64
```

The `nan` was propagated forward from the first value. This is probably not too hard to replicate if we have the right values laying around. 

The problem is that the way the prefix sum works that allows this to be computed in parallel causes `nan`s to propagate throughout the calculation and we will end up with a `nan` everywhere forward of the position where we encounter the first one. Unless anyone has any better ideas, I think they need to be removed from the prefix sum before it runs. cc @harrism who is the expert on this I believe. 

The issue at that point ends up being if we do filter the `nan`s out first and run the prefix sum just using the "condensed" version of the sequence, the numbers will end up being wrong because the nans take up "space" and this means a particular EWMA value needs to downweight the elements that contribute to it appropriately. The 1 in the above sequence needs to be weighted as if it is two positions behind the 3 rather than just one position as it would be in the condensed series (`[1,3]`).

I think mathematically there might be a clever way of inserting constants that fix this. But before I probe further there I thought I would provide an update on my thinking here. Moreover I am curious what libcudf maintainers think is the correct design here given that cudf python might only be one consumer - I hesitate to weave a somewhat pandas specific nan handling strategy fundamentally into the algorithm. 

Curious what others think.

jrhemstad commented:
@brandon-b-miller will this be done for 21.10 burndown or should it be moved?

brandon-b-miller commented:
@jrhemstad close to call - it should be ready for review in the next day or two. I don't have a ton of experience with libcudf so it's a bit hard for me to tell how far off this is from really being airtight. I'm guessing there's a lot to do here, but let me push my latest changes now, and perhaps you will be able to help make that call with a cursory reading.

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+
+/**
+ * @brief modify the source pairs that eventually yield the numerator
+ * and denoninator to account for nan values. Pairs at nan indicies
```
```suggestion
 * and denominator to account for nan values. Pairs at nan indices
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+
+  // Denominator
+  // Fill with pairs
+  thrust::fill(
+    rmm::exec_policy(stream), pairs.begin(), pairs.end(), thrust::pair<double, double>(beta, 1.0));
+
+  if (input.has_nulls()) { pair_beta_adjust(input, pairs, beta, stream); }
```
Filling the pair vector with a constant value seems wasteful. If `pair_beta_adjust` returned its output and took an iterator as input instead of a vector, you could pass a `constant_iterator` as input to this call to `pair_beta_adjust` and avoid the fill.  (In general when I see calls to `thrust::fill` or `thrust::sequence` I look for more efficient approaches, often involving iterators)

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                           rmm::cuda_stream_view stream,
+                           rmm::mr::device_memory_resource* mr)
+{
+  rmm::device_uvector<thrust::pair<double, double>> pairs(input.size(), stream, mr);
```
Temporary data uses default MR.

```suggestion
  rmm::device_uvector<pair_type> pairs(input.size(), stream);
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                    input.end<double>(),
+                    pairs.begin(),
+                    [=] __host__ __device__(double input) -> thrust::pair<double, double> {
+                      return thrust::pair<double, double>(beta, (1.0 - beta) * input);
```
```suggestion
                      return {beta, (1.0 - beta) * input};
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                    input.begin<double>(),
+                    input.end<double>(),
+                    pairs.begin(),
+                    [=] __host__ __device__(double input) -> thrust::pair<double, double> {
```
```suggestion
                    [=] __device__(double input) -> pair_type {
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+    pairs.end(),
+    output.begin<double>(),
+    output.begin<double>(),
+    [=] __host__ __device__(thrust::pair<double, double> pair, double numerator) -> double {
```
```suggestion
    [=] __device__(thrust::pair<double, double> pair, double numerator) -> double {
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                    pairs.begin(),
+                    pairs.end(),
+                    output.begin<double>(),
+                    [=] __host__ __device__(thrust::pair<double, double> pair) -> double {
```
```suggestion
                    [=] __device__(thrust::pair<double, double> pair) -> double {
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+
+  thrust::transform(rmm::exec_policy(stream),
+                    input.begin<double>(),
+                    input.begin<double>() + 1,
```
```suggestion
                    std::next(input.begin<double>()),
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                    input.begin<double>(),
+                    input.begin<double>() + 1,
+                    pairs.begin(),
+                    [=] __host__ __device__(double input) -> thrust::pair<double, double> {
+                      return thrust::pair<double, double>(beta, input);
+                    });
```
```suggestion
                    [=] __device__(double input) {
                      return {beta, input};
                    });
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                    pairs.begin(),
+                    pairs.end(),
+                    output.begin<double>(),
+                    [=] __host__ __device__(thrust::pair<double, double> pair) -> double {
```
```suggestion
                    [=] __device__(pair_type pair) {
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+}
+
+std::unique_ptr<column> ewma(column_view const& input,
+                             double com,
```
What is `com`? Please spell it out.

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+  auto means_view               = means.get()[0].mutable_view();
+
+  // construct deltas: (x_i - mu_i)(x_i - mu_{i-1})
+  rmm::device_uvector<double> d_this(input.size(), stream, mr);
+  rmm::device_uvector<double> d_last(input.size(), stream, mr);
```
I think these are temporary data, so don't pass mr.

```suggestion
  rmm::device_uvector<double> d_this(input.size(), stream);
  rmm::device_uvector<double> d_last(input.size(), stream);
```

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                    input.begin<double>() + 1,
+                    input.end<double>(),
+                    means_view.begin<double>() + 1,
+                    d_this.begin() + 1,
```
Note you are not writing to `d_this[0]` here, and since it is a `device_uvector`, it is uninitialized.

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                    input.begin<double>() + 1,
+                    input.end<double>(),
+                    means_view.begin<double>(),
+                    d_last.begin() + 1,
```
`d_last[0]` is uninitialized too.

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                    thrust::minus<double>());
+
+  // (x_i - mu_i)(x_i - mu_{i-1})
+  thrust::transform(rmm::exec_policy(stream),
+                    d_this.begin(),
+                    d_this.end(),
+                    d_last.begin(),
+                    means_view.begin<double>(),
+                    thrust::multiplies<double>());
```
Here you are reading uninitialized `d_this[0]` and `d_last[0]`, so this is UB.

I think you need to add

```
d_this.set_element_to_zero_async(0, stream);
d_last.set_element_to_zero_async(0, stream);
``` 

before this final call  to `thrust::transform`.

harrism commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                    var_view.begin<double>(),
+                    var_view.end<double>(),
+                    var_view.begin<double>(),
+                    [=] __host__ __device__(double input) -> double { return sqrt(input); });
```
```suggestion
                    [=] __device__(double input) { return sqrt(input); });
```

harrism reviewed (changes requested):
Impressive PR! Lots of C++ simplifications / tidying. At a high level, I'm a little concerned that `double` is hard-coded everywhere. What if `float` columns are input? `double` is twice the storage and bandwidth than `float`.  What if we add support for fp16 in the future?

C++ function documentation is a bit lacking, and some variable names could be improved for clarity. Also, the need for `__host__ __device__` is pretty rare -- definitely not needed here.

brandon-b-miller commented:
@harrism thank you for the review! Hopefully I can template over some of this machinery - things are only `double` because it felt like a good place to start to get the algorithm sorted out. I will make sure to address these items.

harrism commented:
I only noticed this was draft after I did the review!

brandon-b-miller commented:
> I only noticed this was draft after I did the review!

it's never too early for many of these suggestions, and help tidying and thinking through things as I go is always welcome :)

More to come here.

jrhemstad commented on cpp/src/reductions/scan/scan.cuh:
```diff
@@ -89,22 +76,11 @@ std::unique_ptr<column> scan_agg_dispatch(const column_view& input,
-      bool adjust = (dynamic_cast<ewmvar_aggregation*>(agg.get()))->adjust;
-      return ewmvar(input, com, adjust, stream, mr);
-    }
+    case aggregation::EWMA:
+    case aggregation::EWMVAR:
```
```suggestion
    case aggregation::EWMA: [[fallthrough]];
    case aggregation::EWMVAR: [[fallthrough]];
```

brandon-b-miller commented:
Just a quick update on this PR as it has changed significantly. For now I have removed `ewmvar` and `ewmstd` in favor of working through just `ewma` first as there is plenty to review/discuss just with that. It also turns out that `ewmvar` has up to 8 different cases to support - it needs to support `adjust` with and without nulls (4 cases) just like `ewma`, and also `bias=True/False`. Each of these 8 cases arguably needs a slightly different code path and not all of them need the same amount of scratch space to get to the right answer, so I am of the opinion they should be added one by one. 

I added a null handling strategy that ends up giving us the same results that pandas gets for `nan` values. Last I checked `ewm` was not supported for nullable dtypes in pandas, not sure if this is changed. This turned out to be really tricky to do in the context of the pairwise prefix sum this algorithm revolves around, but after a lot of fiddling I managed to figure out that you can sprinkle constants in the right spots to hack it so that we get what we want in the end. More details can be found in the docstring for `make_ewma_aggregation` and the math I ended up with is in this notebook:

https://gist.github.com/brandon-b-miller/ddf8e8cbb4ea9204bc2a01cba0c0ed3a

A couple of files have moved around. I tried to address @harrism 's comments on the original diff as I ported things over. I think this is getting a lot closer to how I had originally envisioned it and should be ready for review soon.

brandon-b-miller commented:
I have this in a position now where I believe it is ready for an initial round of feedback.

vyasr commented on cpp/include/cudf/aggregation.hpp:
```diff
@@ -276,6 +277,12 @@ std::unique_ptr<Base> make_nth_element_aggregation(
+ * @brief Factory to create a EWMA aggregation
+ */
+template <typename Base = aggregation>
+std::unique_ptr<Base> make_ewma_aggregation(double com, bool adjust);
```
```suggestion
std::unique_ptr<Base> make_ewma_aggregation(double const com, bool const adjust);
```

vyasr commented on cpp/include/cudf/detail/aggregation/aggregation.hpp:
```diff
@@ -604,6 +607,41 @@ class row_number_aggregation final : public rolling_aggregation {
+ */
+class ewma_aggregation final : public rolling_aggregation {
+ public:
+  double com;
+  bool adjust;
```
```suggestion
  double const com;
  bool const adjust;
```

vyasr commented on cpp/include/cudf/detail/aggregation/aggregation.hpp:
```diff
@@ -604,6 +607,41 @@ class row_number_aggregation final : public rolling_aggregation {
+  double com;
+  bool adjust;
+
+  ewma_aggregation(double com, bool adjust) : aggregation{EWMA}
+  {
+    this->com    = com;
+    this->adjust = adjust;
+  }
```
```suggestion
  ewma_aggregation(double const com, bool const adjust) : aggregation{EWMA}, com{com}, adjust{adjust} {}
```

vyasr commented on cpp/src/aggregation/aggregation.cpp:
```diff
@@ -596,6 +607,14 @@ std::unique_ptr<Base> make_row_number_aggregation()
 
+/// Factory to create an EWMA aggregation
+template <typename Base>
+std::unique_ptr<Base> make_ewma_aggregation(double com, bool adjust)
```
```suggestion
std::unique_ptr<Base> make_ewma_aggregation(double const com, bool const adjust)
```

vyasr commented on cpp/src/aggregation/aggregation.cpp:
```diff
@@ -596,6 +607,14 @@ std::unique_ptr<Base> make_row_number_aggregation()
+{
+  return std::make_unique<detail::ewma_aggregation>(com, adjust);
+}
+template std::unique_ptr<aggregation> make_ewma_aggregation<aggregation>(double com, bool adjust);
```
```suggestion
template std::unique_ptr<aggregation> make_ewma_aggregation<aggregation>(double const com, bool const adjust);
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+namespace detail {
+
+template <typename T>
+class recurrence_functor {
```
Maybe `term_functor` would be more appropriate since it generates the terms of the recurrence, not the whole recurrence relation?

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+
+template <typename T>
+class recurrence_functor {
+  using pair_type = thrust::pair<T, T>;
```
Rather than defining this in this class, it might be worth defining an alias at the namespace level since these pairs are present everywhere.
```
template <typename T>
using pair_type = thrust::pair<T, T>
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+void compute_recurrence(rmm::device_uvector<thrust::pair<T, T>>& input,
+                        rmm::cuda_stream_view stream)
+{
+  recurrence_functor<T> op;
+  thrust::inclusive_scan(rmm::exec_policy(stream), input.begin(), input.end(), input.begin(), op);
```
```suggestion
  thrust::inclusive_scan(rmm::exec_policy(stream), input.begin(), input.end(), input.begin(), recurrence_functor<T>{});
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+  rmm::device_uvector<cudf::size_type> output(
+    input.size(), stream, rmm::mr::get_current_device_resource());
+
+  auto device_view = *column_device_view::create(input);
```
Don't discard the unique pointer. You should store it as is, then dereference it wherever you need it.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+  auto device_view = *column_device_view::create(input);
+  auto valid_it    = cudf::detail::make_validity_iterator(device_view);
+
+  // TODO - not sure why two iterators produce a different result
```
Is this still relevant?

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                    valid_it,
+                    valid_it + input.size(),
+                    output.begin(),
+                    [=] __device__(bool valid) -> bool { return 1 - valid; });
```
```suggestion
                    [] __device__(bool valid) -> bool { return 1 - valid; });
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+
+/**
+* @brief modify the source pairs that eventually yield the numerator
+* and denoninator to account for nan values. Pairs at nan indicies
```
```suggestion
* and denominator to account for nan values. Pairs at nan indices
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                      rmm::cuda_stream_view stream)
+{
+  using pair_type        = thrust::pair<T, T>;
+  auto device_view       = *column_device_view::create(input);
```
Store the unique pointer and dereference later

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                      rmm::device_uvector<cudf::size_type>& nullcnt,
+                      rmm::cuda_stream_view stream)
+{
+  using pair_type        = thrust::pair<T, T>;
```
Consider a namespace-level alias as suggested above.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+    valid_and_nullcnt + input.size(),
+    pairs.begin(),
+    pairs.begin(),
+    [=] __device__(thrust::tuple<bool, int> valid_and_nullcnt, pair_type pair) -> pair_type {
```
I don't think there's anything that you need to capture here, but ignore this if I'm missing something (although it's probably better to specify exactly which captures you need if I am).
```suggestion
    [] __device__(thrust::tuple<bool, int> valid_and_nullcnt, pair_type pair) -> pair_type {
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+    pairs.begin(),
+    pairs.begin(),
+    [=] __device__(thrust::tuple<bool, int> valid_and_nullcnt, pair_type pair) -> pair_type {
+      bool valid = thrust::get<0>(valid_and_nullcnt);
+      int exp    = thrust::get<1>(valid_and_nullcnt);
```
```suggestion
      bool const valid = thrust::get<0>(valid_and_nullcnt);
      int const exp    = thrust::get<1>(valid_and_nullcnt);
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+
+template <typename T>
+rmm::device_uvector<T> compute_ewma_adjust(column_view const& input,
+                                           T beta,
```
```suggestion
                                           T const beta,
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+      pairs.end(),
+      output.begin(),
+      output.begin(),
+      [=] __device__(pair_type pair, T numerator) -> T { return numerator / pair.second; });
```
```suggestion
      [] __device__(pair_type pair, T numerator) -> T { return numerator / pair.second; });
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                      input.begin<T>(),
+                      input.end<T>(),
+                      pairs.begin(),
+                      [=] __device__(T input) -> pair_type {
```
```suggestion
                      [] __device__(T input) -> pair_type {
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                      pairs.begin(),
+                      pairs.end(),
+                      output.begin(),
+                      [=] __device__(pair_type pair) -> T { return pair.second; });
```
```suggestion
                      [] __device__(pair_type pair) -> T { return pair.second; });
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+      pairs.end(),
+      output.begin(),
+      output.begin(),
+      [=] __device__(pair_type pair, T numerator) -> T { return numerator / pair.second; });
```
```suggestion
      [] __device__(pair_type pair, T numerator) -> T { return numerator / pair.second; });
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+
+template <typename T>
+rmm::device_uvector<T> compute_ewma_noadjust(column_view const& input,
+                                             T beta,
```
```suggestion
                                             T const beta,
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                      return {beta, (1.0 - beta) * input};
+                    });
+
+  // TODO: the first pair is WRONG using the above. Reset just that pair
```
Still relevant?

brandon-b-miller replied:
I was wondering here if there was a way of just setting one element (the first one) to be `{beta, input}` rather than having to use a whole transform?

vyasr replied:
A slightly nicer (but still not amazing) solution would be to add a counting iterator to the above transform and condition the return value on whether that input value is equal to 0. It adds some overhead to the evaluation, but avoiding the extra kernel launch here should make that approach always win since you'll only see divergence for warp 0.

I can't think of anything cleaner than that unfortunately.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                    input.begin<T>(),
+                    input.end<T>(),
+                    pairs.begin(),
+                    [=] __device__(T input) -> pair_type {
```
```suggestion
                    [beta] __device__(T input) -> pair_type {
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                    input.begin<T>(),
+                    std::next(input.begin<T>()),
+                    pairs.begin(),
+                    [=] __device__(T input) -> pair_type {
```
```suggestion
                    [beta] __device__(T input) -> pair_type {
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                        }
+                      });
+
+    auto device_view = *column_device_view::create(input);
```
Save the unique pointer.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+      pairs.begin(),
+      pairs.begin(),
+      [=] __device__(thrust::tuple<bool, T> null_and_null_count, pair_type pair) -> pair_type {
+        bool is_valid = thrust::get<0>(null_and_null_count);
+        T factor      = thrust::get<1>(null_and_null_count);
+
+        T ci = pair.first;
+        T cj = pair.second;
```
```suggestion
        bool const is_valid = thrust::get<0>(null_and_null_count);
        T const factor      = thrust::get<1>(null_and_null_count);

        T const ci = pair.first;
        T const cj = pair.second;
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+      null_and_null_count + input.size(),
+      pairs.begin(),
+      pairs.begin(),
+      [=] __device__(thrust::tuple<bool, T> null_and_null_count, pair_type pair) -> pair_type {
```
```suggestion
      [] __device__(thrust::tuple<bool, T> null_and_null_count, pair_type pair) -> pair_type {
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                    pairs.begin(),
+                    pairs.end(),
+                    output.begin(),
+                    [=] __device__(pair_type pair) -> T { return pair.second; });
```
```suggestion
                    [] __device__(pair_type pair) -> T { return pair.second; });
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+                             rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(cudf::is_floating_point(input.type()), "Column must be floating point type");
+  bool adjust = (dynamic_cast<ewma_aggregation*>(agg.get()))->adjust;
+  T com       = (dynamic_cast<ewma_aggregation*>(agg.get()))->com;
```
This seems a little off. are we guaranteed that the provided agg will be of type `ewma_aggregation`, or is it possible for this to be called with an erroneous type? If we're sure it's the right type we can just use a `static_cast`. If we're not, then we should verify that the result of the `dynamic_cast` is not a null pointer.

vyasr replied:
This still needs to be addressed.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+  } else {
+    data = compute_ewma_noadjust(input, beta, stream, mr);
+  }
+  auto col = std::make_unique<column>(
+    cudf::data_type(cudf::type_to_id<T>()), input.size(), std::move(data.release()));
+  return col;
```
```suggestion
  return std::make_unique<column>(
    cudf::data_type(cudf::type_to_id<T>()), input.size(), std::move(data.release()));
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,377 @@
+  return col;
+}
+
+struct ewma_functor {
+  template <typename T>
+  std::unique_ptr<column> operator()(std::unique_ptr<aggregation> const& agg,
+                                     column_view const& input,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+  {
+    return std::unique_ptr<column>();
+  }
+};
+
+template <>
+std::unique_ptr<column> ewma_functor::operator()<float>(std::unique_ptr<aggregation> const& agg,
+                                                        column_view const& input,
+                                                        rmm::cuda_stream_view stream,
+                                                        rmm::mr::device_memory_resource* mr)
+{
+  return ewma<float>(agg, input, stream, mr);
+}
+
+template <>
+std::unique_ptr<column> ewma_functor::operator()<double>(std::unique_ptr<aggregation> const& agg,
+                                                         column_view const& input,
+                                                         rmm::cuda_stream_view stream,
+                                                         rmm::mr::device_memory_resource* mr)
+{
+  return ewma<double>(agg, input, stream, mr);
+}
```
FYI you can put the specializations directly into the class rather than defining them out of band like this. However, I'm more confused about why the default is to return a default-initialized unique pointer to a column. Should that be an error case?

vyasr commented on cpp/src/reductions/scan/scan.cuh:
```diff
@@ -22,8 +22,8 @@
 #include <cudf/utilities/error.hpp>
 #include <cudf/utilities/type_dispatcher.hpp>
 
+#include <iostream>
```
Remove.

vyasr reviewed (changes requested):
Got through one round of almost entirely superficial review. I'll come back for another round, but for now there are a few things to address here. Could you also fix the style tests?

cwharris commented on cpp/include/cudf/aggregation.hpp:
```diff
@@ -286,6 +287,40 @@ std::unique_ptr<Base> make_nth_element_aggregation(
+ * whose values are the exponentially weighted moving average of the input
+ * sequence. Let these values be known as the y_i.
+ *
+ * EWMA aggregations are parameterized by a center of mass (`com`) which
```
Is "center of mass" standard terminology when talking about EWMAs?

brandon-b-miller replied:
It really depends on the treatment/text. Pandas allows you to specify the same number via a bunch of different scalar transformations with different physical interpretations, but ultimately they are all funneled into one (`com`) and then the formulas are implemented in terms of that one. Since that funnel exists as an importable public pandas function, it made sense to me to roll with that one.

vyasr replied:
Yeah now that I've finished the review I see why you chose COM, it's because of the `get_center_of_mass` method in `pandas`. I'm fine sticking with COM if there's no obvious reason to prefer another.

cwharris commented on cpp/include/cudf/aggregation.hpp:
```diff
@@ -286,6 +287,40 @@ std::unique_ptr<Base> make_nth_element_aggregation(
+ * affects the contribution of the previous values (y_{i-1} ... y_0) in
+ * computing the y_i.
+ *
+ * EWMA aggregations are also parameterized by a boolean `adjust`.
+ * Special considerations have to be given to the mathematical treatment of
+ * the first value of the input sequence. There are two approaches to this,
+ * one which considers the first value of the sequence to be the expoential
+ * weighted moving average of some infinite history of data, and one which
+ * takes the first value to be the only datapoint known. These assumptions
+ * lead to two different formulas for the y_i. `adjust` selects which one.
```
while two values are adequate to communicate which formula to use, I think it would be prudent to use an enum class here with relevant names for each enum value. Not sure what the names would be, though...

https://github.com/pandas-dev/pandas/issues/8861

brandon-b-miller replied:
👍 I think I remember it saying something about this in the developer guide as well. How about `INFINITE_HISTORY` and `FINITE_HISTORY`?

vyasr replied:
What would you name the enum itself? I don't particularly like
```
enum class adjust {
  FINITE_HISTORY,
  INFINITE_HISTORY
};
```
The name `adjust` doesn't really tell you anything about what it means. Maybe
```
enum class HISTORY {
  FINITE,
  INFINITE
}
```
I can't think of any other options here. Even if further levers are added to the EWM calculation I think they'll be orthogonal to this one, so this naming suggestion doesn't feel that constraining to me.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,384 @@
+    As such, the pairs must be updated before summing like the adjusted case to
+    properly downweight the previous values. But now but we also need to compute
+    the normalization factors and divide the results into them at the end.
+
+    */
```
```suggestion
    */
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,384 @@
+    y_{i+1} = (1 - alpha)x_{i-1} + alpha x_i, but really there is a "denominator"
+    which is the sum of the weights: alpha + (1 - alpha) == 1. If a null is
+    encountered, that means that the "previous" value is downweighted by a
+    factor (for each missing value). For example this would y_2 be for one NULL:
```
You're not capitalizing null anywhere else that I saw, no reason to start now.
```suggestion
    factor (for each missing value). For example this would y_2 be for one null:
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,384 @@
+  T com       = (dynamic_cast<ewma_aggregation*>(agg.get()))->com;
+  // center of mass is easier for the user, but the recurrences are
+  // better expressed in terms of the derived parameter `beta`
+  T beta = 1.0 - (1.0 / (com + 1.0));
```
FWIW this is equivalent to 
```suggestion
  T beta = com / (com + 1.0);
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,384 @@
+  // better expressed in terms of the derived parameter `beta`
+  T beta = 1.0 - (1.0 / (com + 1.0));
+
+  //rmm::device_uvector<T> data(input.size(), stream, mr);
+  if (adjust) {
+    return std::make_unique<column>(
+      cudf::data_type(
+        cudf::type_to_id<T>()), 
+        input.size(), 
+        std::move(
+          compute_ewma_adjust(input, beta, stream, mr).release()
+        )
+    );
+  } else {
+    return std::make_unique<column>(
+      cudf::data_type(
+        cudf::type_to_id<T>()), 
+        input.size(), 
+        std::move(
+          compute_ewma_noadjust(input, beta, stream, mr).release()
+        )
+    );
+  }
```
```suggestion
  auto result = [&] () {
    if (adjust) {
      return compute_ewma_adjust(input, beta, stream, mr).release();
    } else {
      return compute_ewma_noadjust(input, beta, stream, mr).release();
    }
  }();
  return std::make_unique<column>(
    cudf::data_type(cudf::type_to_id<T>()), 
    input.size(), 
    std::move(result)
  );
```

vyasr commented on cpp/src/reductions/scan/scan.cuh:
```diff
@@ -33,6 +32,11 @@ rmm::device_buffer mask_scan(column_view const& input_view,
                              rmm::cuda_stream_view stream,
                              rmm::mr::device_memory_resource* mr);
 
+std::unique_ptr<column> ewm(column_view const& input,
```
Why is this named `ewm` when everything else is named `ewma`?

brandon-b-miller replied:
Ah, this originally dispatched to all the other weighted moments (`ewmvar` and `ewmstd`) which I removed to reduce the complexity here. Would it be better to just remove it for now and then put it back in when those functions come along?

vyasr commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -275,7 +275,9 @@ std::unique_ptr<column> scan_inclusive(
 
-  if (null_handling == null_policy::EXCLUDE) {
+  if (agg->kind == aggregation::EWMA) {
+    return output;
```
Is EWMA the only aggregation that's doing its own null handling (because it has special logic)? Is that why this special case exists here?

brandon-b-miller replied:
That's correct. I got the idea from how `RANK` worked at the time I started implementing this. Happy to change it to work differently if theres an obvious way that would be better, but I am a little fuzzier on this area of the pipeline in general so it's harder to tell if something is a good idea or not.

cwharris replied:
Is null handling a requirement? If there's no precedence here, we should wait until we have a use case. It seems odd that `mean` would handle nulls when we suspect the others will not or cannot.

brandon-b-miller replied:
I believe all of the other scan functions handle nulls, its just that a lot of the null handling is common to the scan functions (min, max, prod, sum) so it can be abstracted out. Rank and ewm have to do their own thing.

vyasr commented on python/cudf/cudf/core/window/ewm.py:
```diff
@@ -0,0 +1,143 @@
+        raise NotImplementedError("ewmstd not yet supported.")
+
+    def corr(self, other):
+        raise NotImplementedError("corr not yet supported.")
```
The above errors prefix the method name with `ewm` (e.g. `ewmvar not yet supported`). I personally prefer not having the prefix, but I don't really care as long as you're consistent.

vyasr commented on python/cudf/cudf/core/window/ewm.py:
```diff
@@ -0,0 +1,143 @@
+        if not is_numeric_dtype(sr.dtype):
+            raise TypeError("No numeric types to aggregate")
+
+        kws = {
+            "com": self.com,
+            "adjust": self.adjust,
+        }
```
This seems unnecessary. Just pass the kwargs directly to the `scan` call.

vyasr reviewed (changes requested):
In addition to my inline comments, I think you should look to fuse a few more of the kernel invocations. Inlining more of the logic in `ewm.cu` into fewer (albeit more complex) lambdas for use in transforms would probably help coalesce memory accesses. In particular, can't the `pair_beta_adjust` logic be part of the transform that computes the original pairs based on `input.has_nulls()`? I would also look into `thrust::transform_inclusive_scan`, which would allow you to fuse the scan as well.

brandon-b-miller commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,384 @@
+};
+
+/**
+ * @brief Solve a recurrence relation using a blelloch scan
```
Added this

brandon-b-miller commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,384 @@
+ * The second elements of the pairs will contain the result
+ */
+template <typename T>
+void compute_recurrence(rmm::device_uvector<pair_type<T>>& input, rmm::cuda_stream_view stream)
```
Not anymore. I think this used to be a little more complicated but I cut it down a lot over time. I went ahead and inlined this

brandon-b-miller commented:
> In addition to my inline comments, I think you should look to fuse a few more of the kernel invocations. Inlining more of the logic in `ewm.cu` into fewer (albeit more complex) lambdas for use in transforms would probably help coalesce memory accesses. In particular, can't the `pair_beta_adjust` logic be part of the transform that computes the original pairs based on `input.has_nulls()`? I would also look into `thrust::transform_inclusive_scan`, which would allow you to fuse the scan as well.

I managed to completely factor out `pair_beta_adjust` and move its logic to the data generation stage of the algorithm. As for `transform_inclusive_scan`, stay tuned...

brandon-b-miller commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,424 @@
+};
+
+template <typename T>
+class ewma_adjust_numerator_nonull_functor : public ewma_functor_base<T> {
```
these names don't seem great, what should I call these?

vyasr replied:
Before addressing naming, let's see if there's a better approach here. I think you can implement all of these functors as a single functor with a couple of additional template parameters and suitable use of `if constexpr`. Could you give that a shot? For example, you could do things like
```
// We may want to make has_nulls a runtime nullate rather than a template parameter, but for now let's just try it with templating since that's easier.
template <typename T, bool adjust, bool has_nulls, bool is_numerator>
...
auto input = [=] () {
    if constexpr (is_numerator) {
        return thrust::get<2>(data);
    } else {
        return 1.0;
    }
}();
...
```
That's specifically an example of using the template parameter to make a compile-time decision on what the second value of the pair should be based on whether we're computing the numerator or the denominator, but I think you could do similar things for handling adjustment and nullability. I'm not sure that you'll be able to combine _all_ of the functors that way, and some of them may just be clearer/more readable if you leave them with separate implementations, but right now there's a lot of code duplication and the naming confusion is partly a result of that. Let's try and use templates to resolve this at least part way.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,424 @@
+
+template <typename T>
+class ewma_adjust_numerator_nonull_functor : public ewma_functor_base<T> {
+ private:
+  T beta;
+
+ public:
+  ewma_adjust_numerator_nonull_functor(T beta) : beta{beta} {}
```
You can define `beta` and the constructor in `ewma_functor_base<T>`, then every child class can just write
```
using ewma_functor_base<T>::ewma_functor_base;
using ewma_functor_base<T>::beta;
```
to have access to the constructor and `beta` (without having to do `this->beta`).

vyasr replied:
Actually put this on hold in favor of trying my `if constexpr` solution above.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,424 @@
+
+  // center of mass is easier for the user, but the recurrences are
+  // better expressed in terms of the derived parameter `beta`
+  T beta = com / (com + 1.0);
```
```suggestion
  T const beta = com / (com + 1.0);
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,424 @@
+  CUDF_EXPECTS(cudf::is_floating_point(input.type()), "Column must be floating point type");
+
+  cudf::ewm_history history = ewma_agg->history;
+  T com                     = ewma_agg->com;
```
```suggestion
  T const com                     = ewma_agg->com;
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,424 @@
+  if (ewma_agg == NULL) { CUDF_FAIL("Expected an EWMA aggregation."); }
+  CUDF_EXPECTS(cudf::is_floating_point(input.type()), "Column must be floating point type");
+
+  cudf::ewm_history history = ewma_agg->history;
```
```suggestion
  cudf::ewm_history const history = ewma_agg->history;
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,424 @@
+  return ewma<double>(agg, input, stream, mr);
+}
+
+std::unique_ptr<column> ewm(column_view const& input,
```
This should probably be `ewma` instead of `ewm` right?

vyasr commented on python/cudf/cudf/core/window/ewm.py:
```diff
@@ -0,0 +1,140 @@
+
+    Returns
+    -------
+    ``ExponentailMovingWindow`` object
```
```suggestion
    ``ExponentialMovingWindow`` object
```

vyasr replied:
Although the proper style for returns once we start linting them is `type : description`, e.g. `ExponentialMovingWindow : An object providing various EWM functions`.

vyasr reviewed (changes requested):
Do we have any benchmarks of how this performs? I'd also be curious whether any of the changes that have been made over the rounds of review have had a measurable impact on performance.

Unless it's urgent, I'd suggest moving this PR to 22.04. I think it'll require at least one more round of review, and there are additional reviewers who need to provide input as well and they haven't looked at this in a while.

brandon-b-miller commented:
> Do we have any benchmarks of how this performs? I'd also be curious whether any of the changes that have been made over the rounds of review have had a measurable impact on performance.
> 
> Unless it's urgent, I'd suggest moving this PR to 22.04. I think it'll require at least one more round of review, and there are additional reviewers who need to provide input as well and they haven't looked at this in a while.

I do have some initial benchmarking. Here is the gist. It looks like the breakeven point in terms of data size is usually in the tens of thousands of elements and it goes up from there.  https://gist.github.com/brandon-b-miller/2861d13ddd1731c7285da4747f14bfd2

Retargeting for 22.04.

vyasr commented:
> I do have some initial benchmarking. Here is the gist. It looks like the breakeven point in terms of data size is usually in the tens of thousands of elements and it goes up from there. https://gist.github.com/brandon-b-miller/2861d13ddd1731c7285da4747f14bfd2
> 
> Retargeting for 22.04.

Nice. Do you have any idea what this looked like in earlier iterations of this PR?

brandon-b-miller commented:
> > I do have some initial benchmarking. Here is the gist. It looks like the breakeven point in terms of data size is usually in the tens of thousands of elements and it goes up from there. https://gist.github.com/brandon-b-miller/2861d13ddd1731c7285da4747f14bfd2
> > Retargeting for 22.04.
> 
> Nice. Do you have any idea what this looked like in earlier iterations of this PR?

It has improved for sure, the GPU always won for large data sizes but initially there was a lot of scratch space being used for intermediate data, and the memory allocations for that were definitely a bottleneck.

harrism commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,424 @@
+                    output.begin(),
+                    [] __device__(bool valid) -> bool { return 1 - valid; });
+
+  // 0, 1, 0, 1, 1, 0 -> 0, 0, 1, 0, 0, 2
```
I don't think inclusive_scan_by_key is doing what this comment says since you are using the nulls as the keys and the values:
// keys: 0, 1, 0, 1, 1, 0
// values: 0, 1, 0, 1, 1, 0
// result: 0, 1, 0, 1, 2, 0

harrism commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,424 @@
+  auto device_view = column_device_view::create(input);
+  auto valid_it    = cudf::detail::make_validity_iterator(*device_view);
+
+  // Invert the null iterator
+  thrust::transform(rmm::exec_policy(stream),
+                    valid_it,
+                    valid_it + input.size(),
+                    output.begin(),
+                    [] __device__(bool valid) -> bool { return 1 - valid; });
+
+  // 0, 1, 0, 1, 1, 0 -> 0, 0, 1, 0, 0, 2
+  thrust::inclusive_scan_by_key(
+    rmm::exec_policy(stream), output.begin(), output.end() - 1, output.begin(), output.begin() + 1);
```
Combine these into a scan with inversion applied as a transform on the iterator. However, I note you shift the output by one, which means you leave the first element as its null value (0 or 1). But this means that this first value doesn't influence the scan. Is that intentional?  If it is, my suggestion will need modification.

Also note using `not valid` instead of `1 - valid` to reduce unneeded integer casting.

```suggestion
  // Invert the validity iterator to get nulls
  auto nulls = thrust::make_transform_iterator(valid_it, [] __device__(bool valid) { return not valid; });

  // Scan the nulls 
  // 0, 1, 0, 1, 1, 0 -> 0, 1, 0, 1, 2, 0
  thrust::inclusive_scan_by_key(rmm::exec_policy(stream), nulls, nulls + input.size(), nulls, output.begin());
```

brandon-b-miller replied:
I am having a little trouble getting this to compile, but let me address your question about the first value in the mean time. The reason the output is shifted is because I want to be able to line each value up with the number of nulls that came before it. But since this is the first value, there can't be any nulls that come before it, so this number will always be zero, except in the case where the first value is itself null. Since there didn't seem to be a great way of making sense of a sequence that starts with missing values, I had expected there would be UB in that case.

The only other option I could see would be to find the first non null value and execute the existing algorithm from that point forward in the data, leaving the leading nulls as they are - but I can't convince myself that doesn't lead to some conflicting assumptions (don't we assume the first value is, in fact, the first value ever in the `FINITE` case?)

brandon-b-miller replied:
Just as an update I was never able to produce correct results using this approach. I am fairly (?) sure that this is the logical equivalent:
```
  auto nulls = thrust::make_transform_iterator(valid_it, [] __device__(bool valid) { return not valid; });
  thrust::inclusive_scan_by_key(rmm::exec_policy(stream), nulls, nulls + input.size() - 1, nulls, output.begin() + 1);
```
However this just doesn't produce the same output sequence as the current implementation. I suspect there's something subtle going on with the iterators that escapes me, I tried several permutations including creating separate iterators representing the same null sequence and so forth, but didn't find success. I am leaving this as-is unless @harrism has any insight on what might be going wrong.

vyasr replied:
We resolved this in the latest. The issue with Brandon's snippet above is that the lambda was accepting and outputting booleans. That's fixed now.

harrism commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,424 @@
+                                     recurrence_functor<T>{});
+  }
+
+  // copy the second elements to the output for now
```
Looks like this just gets overwritten below?

brandon-b-miller replied:
I think in the updates since this comment the code that was overwriting this result was removed. At this point I think this is still needed as this stores the result of the numerator's computation into the output memory, freeing up the array of pairs to be reused for the denominator's computation later on.

harrism reviewed (dismissed):
Remember to update copyright to include 2022 on any changed files.

brandon-b-miller commented on cpp/include/cudf/aggregation.hpp:
```diff
@@ -286,6 +288,40 @@ std::unique_ptr<Base> make_nth_element_aggregation(
+ * @param adjust which assumption to make about the first value
+ */
+template <typename Base = aggregation>
+std::unique_ptr<Base> make_ewma_aggregation(double const com, ewm_history history);
```
I am moving towards `center_of_mass` everywhere as you suggest, but I think it should be `ewm_history` still - this is because `ewmvar` and `ewmstd` will still need both options of how to treat the first value and I feel like `ewma_history` refers directly to the moving average version.

brandon-b-miller commented:
Had some time recently to come back to this, thanks everyone for your patience here.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,396 @@
+/**
+ * @brief functor to be summed over in a prefix sum such that
+ * the recurrence in question is solved.
+ * see https://www.cs.cmu.edu/~guyb/papers/Ble93.pdf S. 1.4
```
In case this link ever dies, it would be helpful to include an actual citation of the paper (at least author and title) so that we can find it again.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,396 @@
+};
+
+template <typename T, bool nulls, bool is_numerator>
+class ewma_adjust_functor : public ewma_functor_base<T> {
```
Our workshopping offline has definitely made this significantly cleaner. I'm on board with keeping the adjust and noadjust functors separate, they are definitely different enough.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,396 @@
+
+template <typename T, bool nulls, bool is_numerator>
+class ewma_adjust_functor : public ewma_functor_base<T> {
+ private:
+  T beta;
+
+ public:
+  ewma_adjust_functor(T beta) : beta{beta} {}
```
Both the `beta` declaration and the constructor can be moved to the parent class rather than duplicated in both functors.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,396 @@
+
+  __device__ pair_type<T> operator()(tupletype data)
+  {
+    if constexpr (is_numerator) {
+      if constexpr (nulls) {
```
Try swapping the order of the two conditionals. From looking at the code, the nullable code paths are nearly identical whether we're calculating the numerator or the denominator. You should be able to make a much smaller `if constexpr (is_numerator)` block inside an `if constexpr (nulls)` block.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,396 @@
+
+  using tupletype = std::conditional_t<nulls, thrust::tuple<bool, int, T>, T>;
+
+  __device__ pair_type<T> operator()(tupletype data)
```
The parameter can be const, right?

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,396 @@
+          return {beta, (1.0 - beta) * input};
+        } else if (is_valid and nullcnt != 0) {
+          // one or more preceeding values is null, adjust by how many
+          T factor = (1.0 - beta) + pow(beta, nullcnt + 1);
```
```suggestion
          T const factor = (1.0 - beta) + pow(beta, nullcnt + 1);
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,389 @@
+      int const exp    = thrust::get<1>(data);
+      T const input    = thrust::get<2>(data);
+      T const beta     = this->beta;
+      if constexpr (is_numerator) {
+        if (valid and (exp != 0)) {
+          // The value is non-null, but nulls preceeded it
+          // must adjust the second element of the pair
+
+          return {beta * (pow(beta, exp)), input};
+        } else if (!valid) {
+          // the value is null, carry the previous value forward
+          // "identity operator" is used
+          return this->IDENTITY;
+        } else {
+          return {beta, input};
+        }
+      } else {
+        if (valid and (exp != 0)) {
+          // The value is non-null, but nulls preceeded it
+          // must adjust the second element of the pair
+
+          return {beta * (pow(beta, exp)), 1.0};
+        } else if (!valid) {
+          // the value is null, carry the previous value forward
+          // "identity operator" is used
+          return this->IDENTITY;
+        } else {
+          return {this->beta, 1.0};
+        }
+      }
```
How about
```suggestion
      if (!valid) {
        // The value is null, so carry the previous value forward.
        // Return the identity operand for this operator.
        return this->IDENTITY;
      } else {
        T second = [=] () {
          if constexpr(is_numerator) {
            return input;
          } else {
            return 1;
          }
        }()
        if (valid and (exp != 0)) {
          // The value is non-null, but nulls preceded it so
          // we must adjust the second element of the pair
          return {beta * (pow(beta, exp)), second};
        } else {
          // The exp is 0, so we just propagate the value
          return {beta, second};
        }
      }
```

vyasr replied:
At this point even the IIFE just to use `if constexpr` might be overkill. A simple ternary `T const second = is_numerator ? input : 1` ought to be fine, I'd be surprised if the compiler couldn't figure out how to optimize that just as well as the `if constexpr` and there's no concern about the code compiling.

brandon-b-miller replied:
Refactored this roughly as such

brandon-b-miller commented on cpp/tests/reductions/ewm_tests.cpp:
```diff
@@ -0,0 +1,102 @@
+};
+
+TYPED_TEST_SUITE(TypedEwmScanTest, cudf::test::FloatingPointTypes);
+
```
I am in support of larger tests, however coding up the serial version of the algorithm isn't trivial, especially since we have four different versions of it in this PR. The problem will worsen when `ewmvar` and `ewmstd` are introduced, since those act differently depending on an additional parameter, `bias`. 

In lieu of such would you be ok with perhaps loading some static data generated from pandas? What do you think @harrism ?

vyasr replied:
@harrism what do you think here? IMO this issue is similar to what cuIO faces where in practice it's just much easier to test larger and more complex data from Python, including the cases where we need to test multiple thread blocks etc.

harrism replied:
OK.

vyasr reviewed (changes requested):
@brandon-b-miller apologies I didn't realize I hadn't submitted my reviews. I think we're pretty close now though.

brandon-b-miller commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,379 @@
+{
+  rmm::device_uvector<T> output(input.size(), stream);
+  rmm::device_uvector<pair_type<T>> pairs(input.size(), stream);
+  rmm::device_uvector<cudf::size_type> nullcnt(input.size(), stream);
```
Even if you did this wouldn't you still end up with answers sitting in the second elements of the `pairs` and need to launch a `thrust::copy` to move them to the output array?

brandon-b-miller commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,379 @@
+                                     recurrence_functor<T>{});
+  }
+
+  thrust::transform(
+    rmm::exec_policy(stream),
+    pairs.begin(),
+    pairs.end(),
+    output.begin(),
+    output.begin(),
+    [] __device__(pair_type<T> pair, T numerator) -> T { return numerator / pair.second; });
```
Even if you did this wouldn't you still need to launch a copy to move the data from the second element of each pair into the output array?

vyasr replied:
Why do you ever need to store the intermediate value in the pair though? AFAICT the pairs are output only for the above `transform_inclusive_scan` calls, they aren't used as the input anywhere, so instead of having the functor return the pair couldn't the functor be changed to just return `numerator / pair.second` and write directly into `output`?

vyasr replied:
Note that you might need to zip output into the itr as well, and probably need to double-check that it's safe to read and write that way.

brandon-b-miller replied:
I think I need just a little more help seeing what you are proposing. In this algorithm, we have the following steps
1.  Populate pairs with the correct data to compute the numerators with a prefix sum
2. Perform the prefix sum on the populated pairs
3. Move the second element of the (now prefix-summed) pairs to the output
4. Populate the pairs (reusing the previous allocation) with the correct data to compute the denominators with a prefix sum
5. Perform the prefix sum on the populated pairs
6. Take the second element of the denominators and elementwise divide the numerators located in the output array by them. 

In the current view, (1) and (2) are computed using a transform inclusive scan, after which if we want to reuse that allocation, we need to move the answers out of the memory and into the output (3). At this point (4) and (5) are done together using another transform inclusive scan. (6) is the intended result of this final transform. The pairs necessarily hold the intermediate result of the computation because that's where the 'answers' of the prefix sum land. 

It seems to be you are suggesting a formulation wherein (4), (5), and (6) all happen using one kernel launch. I don't quite follow how we'd do that given the full prefix sum must finish executing before the answers are available to copy to the output. Can you help me out with a quick example or maybe some pseudocode?

vyasr replied:
I think you're right, there is no way to fuse these operations the way that I was hoping because you need the scan to complete before you can do the transform. So never mind!

brandon-b-miller commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,379 @@
+{
+  rmm::device_uvector<T> output(input.size(), stream);
+  rmm::device_uvector<pair_type<T>> pairs(input.size(), stream);
+  rmm::device_uvector<size_type> nullcnt(input.size(), stream);
```
I think this is the same case as above. Here, we just got done doing the prefix sum and the answer is in the second element of our pairs. A copy is needed to move it to the output array.

brandon-b-miller commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,379 @@
+  }
+
+  // copy the second elements to the output for now
+  thrust::transform(rmm::exec_policy(stream),
+                    pairs.begin(),
+                    pairs.end(),
+                    output.begin(),
+                    [] __device__(pair_type<T> pair) -> T { return pair.second; });
+  return output;
```
I think this is the same case as above. Here, we just got done doing the prefix sum and the answer is in the second element of our pairs. A copy is needed to move it to the output array.

brandon-b-miller commented:
@vyasr thanks for circling back to this! I've retargeted to the active branch and brought the feature branch up to date, as well as addressed your comments.

vyasr commented:
@brandon-b-miller you're running into issues because of my breaking changes from #11960. You'll need to update a few function calls, that PR has examples of the change.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,385 @@
+    }
+  }();
+  if (input.has_nulls()) {
+    nullcnt = null_roll_up(input, stream);
```
This is superfluous now.

vyasr replied:
```suggestion
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,385 @@
+    properly downweight the previous values. But now but we also need to compute
+    the normalization factors and divide the results into them at the end.
+    */
+    nullcnt          = null_roll_up(input, stream);
```
```suggestion
```

vyasr reviewed (changes requested):
This is getting pretty close now. There's a compilation issue due to some upstream changes, but once you merge them and tests pass I think you should re-request review from the other reviewers. I think this PR is now in a place where it would benefit from a few additional eyes.

davidwendt commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,385 @@
+  }
+};
+
+std::unique_ptr<column> ewma(column_view const& input,
```
Would it be ok to spell out at least this function name?
```suggestion
std::unique_ptr<column> exponential_weighted_moving_average(column_view const& input,
```

vyasr commented:
@harrism @jrhemstad @cwharris I think this PR is ready for another round of review. There may be a bit more room for simplification still, but it probably needs some fresh eyes other than mine to find them.

@brandon-b-miller aside from my remaining open comments around cleaning up the `null_roll_up` usage I think I'm almost good other than reviewing tests and any cosmetic things I may catch on a final round of review.

harrism commented:
I don't really have time to go in depth on this now. Happy for someone else to take my spot.

bdice commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,382 @@
+      return rmm::device_uvector<cudf::size_type>(input.size(), stream);
+    }
+  }();
+  // denominators are all 1 so dont need to be computed
```
```suggestion
  // denominators are all 1 and do not need to be computed
```

bdice commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,382 @@
+
+struct ewma_functor {
+  template <typename T>
+  std::enable_if_t<!is_floating_point<T>(), std::unique_ptr<column>> operator()(
```
If allowed by the compiler, use SFINAE with `CUDF_ENABLE_IF` here and below.

It's option 3 instead of option 1 from this demo: https://godbolt.org/z/ob4cr99Pb

bdice commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,382 @@
+  }
+};
+
+std::unique_ptr<column> exponential_weighted_moving_average(column_view const& input,
```
I think the term is usually "exponential_ly_ weighted". https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average

Or use "exponential moving average" if you want to go for EMA instead of EWMA everywhere. Anything consistent is fine with me.

```suggestion
std::unique_ptr<column> exponentially_weighted_moving_average(column_view const& input,
```

bdice commented on python/cudf/cudf/_lib/cpp/aggregation.pxd:
```diff
@@ -139,6 +143,10 @@ cdef extern from "cudf/aggregation.hpp" namespace "cudf" nogil:
         string user_defined_aggregator,
         data_type output_type) except +
 
+    cdef unique_ptr[T] make_ewma_aggregation[T](
```
Order this at the end (or consistently with the ordering of other aggregators in other files).

brandon-b-miller commented on cpp/include/cudf/detail/aggregation/aggregation.hpp:
```diff
@@ -631,6 +634,40 @@ class row_number_aggregation final : public rolling_aggregation {
+class ewma_aggregation final : public scan_aggregation {
+ public:
+  double const center_of_mass;
+  cudf::ewm_history history;
```
The issue with this is that we're later going to have `ewmvar` and `ewmstd`, each of which can take either one of the treatments of the first value, so they'll both be dependent on this parameter. `ewm_history` is general all three APIs will return differing results depending on which treatment is applied. `ewma` only refers to the moving average calculation.

bdice replied:
Gotcha. That makes more sense.

neoblackxt commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,382 @@
+                    output.begin(),
+                    [] __device__(bool valid) -> bool { return 1 - valid; });
+
+  // null mask {0, 1, 0, 1, 1, 0} leads to output array {0, 0, 1, 0, 0, 2}
```
```suggestion
  // null mask {0, 1, 0, 1, 1, 0} leads to output array {0, 0, 1, 0, 1, 2}
```

brandon-b-miller commented:
This has been at the bottom of the stack for a while, but having come up in a number of separate conversations recently I'll see what needs to be done to get things merged here.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,382 @@
+  auto valid_it    = cudf::detail::make_validity_iterator(*device_view);
+
+  // Invert the null iterator
+  thrust::transform(rmm::exec_policy(stream),
```
Looks like this is exactly what Mark asked about https://github.com/rapidsai/cudf/pull/9027#discussion_r790346296.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+  pair_type<T> IDENTITY = {1.0, 0.0};
+};
+
+template <typename T, bool nulls, bool is_numerator>
```
Let's use `has_nulls` rather than just `nulls` consistently in this file.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+
+  __device__ pair_type<T> operator()(tupletype const data)
+  {
+    if constexpr (nulls) {
```
It looks like the nulls/non-nulls branches of this operator are completely distinct. Also, the `tupletype` above is conditioned on that. Can we just split these implementations into separate template specializations instead? `if constexpr` seems like an odd choice in this case since the functors are effectively completely different.

brandon-b-miller replied:
Originally, I moved towards this approach of using `if constexpr` for the null case following your suggestion [in a previous round of review](https://github.com/rapidsai/cudf/pull/9027#discussion_r788193523). Since I've managed to miss precisely what you're requesting a couple of times, would you mind helping me out with a commit here @vyasr ? I'm happy to review.

vyasr replied:
Yup, I remember. Sorry for the runaround. Things have changed a bit since then in part because we can use structured bindings now so a lot of the shared bits become trivial. Let me give this a shot.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+  T beta;
+  ewma_functor_base(T beta) : beta{beta} {}
+
+  pair_type<T> IDENTITY = {1.0, 0.0};
```
This should work.
```suggestion
  pair_type<T> IDENTITY{1.0, 0.0};
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+  __device__ pair_type<T> operator()(tupletype const data)
+  {
+    if constexpr (nulls) {
+      bool const valid = thrust::get<0>(data);
+      int const exp    = thrust::get<1>(data);
+      T const input    = thrust::get<2>(data);
```
We can unpack these with a structured binding now I think (works in newer CCCL versions).
```suggestion
      auto const [valid, exp, input] = data;
```

brandon-b-miller replied:
This seems to conflict later on with the way the lambda below is expressed:

```
...Structured binding cannot be captured...
```

vyasr replied:
Ah yeah this is an unfortunate limitation pre-C++20 (search for "lambda" in https://en.cppreference.com/w/cpp/language/structured_binding). We can work around that with an alias.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+    size_type const index = thrust::get<1>(data);
+    T const input         = thrust::get<0>(data);
+
+    if constexpr (!nulls) {
```
As above, I think we should use explicit template specializations instead of `if constexpr` here and we can define the right `tupletype` explicitly. Once you do that you can use the structured bindings I suggested above here as well (without that you'd have to copy the index/input unpacking into both specializations as separate lines, but now it's all a one-liner).

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+    cudf::detail::make_validity_iterator(*device_view),
+    cuda::proclaim_return_type<int>([] __device__(int valid) -> int { return 1 - valid; }));
+
+  // null mask {0, 1, 0, 1, 1, 0} leads to output array {0, 0, 1, 0, 1, 2}
```
Since we're doing the inversion as part of it now
```suggestion
  // valid mask {1, 0, 1, 0, 0, 1} leads to output array {0, 0, 1, 0, 1, 2}
```

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+                                     recurrence_functor<T>{});
+
+  } else {
+    /*
+    In this case, a denominator actually has to be computed. The formula is
+    y_{i+1} = (1 - alpha)x_{i-1} + alpha x_i, but really there is a "denominator"
+    which is the sum of the weights: alpha + (1 - alpha) == 1. If a null is
+    encountered, that means that the "previous" value is downweighted by a
+    factor (for each missing value). For example this would y_2 be for one null:
+    data = {x_0, NULL, x_1},
+    y_2 = (1 - alpha)**2 x_0 + alpha * x_2 / (alpha + (1-alpha)**2)
+
+    As such, the pairs must be updated before summing like the adjusted case to
+    properly downweight the previous values. But now but we also need to compute
+    the normalization factors and divide the results into them at the end.
+    */
```
This comment doesn't seem to be referring to the code below it. Perhaps it should be moved to one of the functors above?

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+                             rmm::device_async_resource_ref mr)
+{
+  auto ewma_agg = (dynamic_cast<ewma_aggregation const*>(&agg));
+  if (ewma_agg == NULL) { CUDF_FAIL("Expected an EWMA aggregation."); }
```
Is this actually reachable? A `CUDF_FAIL` here seems out of place because you would only get to this point if other libcudf code (the dispatcher) incorrectly forwarded the wrong kind of aggregation, right? I think we can remove this line.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+{
+  auto ewma_agg = (dynamic_cast<ewma_aggregation const*>(&agg));
+  if (ewma_agg == NULL) { CUDF_FAIL("Expected an EWMA aggregation."); }
+  CUDF_EXPECTS(cudf::is_floating_point(input.type()), "Column must be floating point type");
```
This is redundant, the SFINAE in the `ewma_functor` should handle it IIUC.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+                                  std::move(result),
+                                  rmm::device_buffer{},
+                                  0);
+}
```
Let's just inline this function in the functor overload below.

vyasr commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,374 @@
+  }();
+  return std::make_unique<column>(cudf::data_type(cudf::type_to_id<T>()),
+                                  input.size(),
+                                  std::move(result),
```
Could you remove the two releases above and instead release once here?

vyasr commented on cpp/src/reductions/scan/scan_inclusive.cu:
```diff
@@ -182,7 +182,8 @@ std::unique_ptr<column> scan_inclusive(column_view const& input,
   auto output = scan_agg_dispatch<scan_dispatcher>(
     input, agg, static_cast<bitmask_type*>(mask.data()), stream, mr);
-  output->set_null_mask(std::move(mask), null_count);
+  // Use the null mask produced by the op for EWM
+  if (agg.kind != aggregation::EWMA) { output->set_null_mask(std::move(mask), null_count); }
```
I'm OK with this, but open question for other reviewers if they would prefer a more systematic approach to handling this. I'm inclined to leave it as-is for now and not try to overgeneralize to a single case.

vyasr commented on cpp/tests/reductions/ewm_tests.cpp:
```diff
@@ -0,0 +1,106 @@
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/reduction.hpp>
+
+using aggregation = cudf::aggregation;
+using cudf::null_policy;
+using cudf::scan_type;
+using namespace cudf::test::iterators;
```
I would just spell these out explicitly below. It's not that much more verbose, and we typically prefer to avoid using declarations for this kind of shortening.

vyasr commented on cpp/tests/reductions/ewm_tests.cpp:
```diff
@@ -0,0 +1,106 @@
+  auto const v = make_vector<TypeParam>({1.0, 2.0, 3.0, 4.0, 5.0});
+  auto col     = this->make_column(v);
+
+  auto const expected_ewma_vals_adjust = cudf::test::fixed_width_column_wrapper<TypeParam>{
+    {1.0, 1.75, 2.61538461538461497469, 3.54999999999999982236, 4.52066115702479365268}};
+
+  auto const expected_ewma_vals_noadjust =
+    cudf::test::fixed_width_column_wrapper<TypeParam>{{1.0,
+                                                       1.66666666666666651864,
+                                                       2.55555555555555535818,
+                                                       3.51851851851851815667,
+                                                       4.50617283950617242283}};
```
These are a lot of digits of precision. Maybe we should consider rounding off some digits and comparing that instead or something? I'm not sure.

vyasr commented on python/cudf/cudf/core/window/ewm.py:
```diff
@@ -0,0 +1,152 @@
+    axis : {0, 1}, default 0
+        Not Supported
+    times : str, np.ndarray, Series, default None
+        Not Supported
```
Can we add a `pandas-compat` doc section indicating the parameters that aren't supported (and any other differences)?

vyasr commented on python/cudf/cudf/core/window/ewm.py:
```diff
@@ -0,0 +1,152 @@
+from typing import Union
+
+import numpy as np
+from pandas.core.window.ewm import get_center_of_mass
```
Is this considered part of the pandas API? @mroeschke is there a way to access this via a more public import (we've been trying to avoid pandas.core)?

mroeschke replied:
I don't think this will be apart of the pandas public API since it's only a very specific helper function. I would just vendor over this function here since it's fairly small

vyasr reviewed (changes requested):
Some smaller things, then hopefully we can wrap this up!

brandon-b-miller commented:
/ok to test

vyasr commented:
/ok to test

brandon-b-miller commented:
/ok to test

bdice commented on cpp/include/cudf/aggregation.hpp:
```diff
@@ -411,6 +414,42 @@ std::unique_ptr<Base> make_nth_element_aggregation(
+ *
+ * @param center_of_mass the center of mass.
+ * @param history which assumption to make about the first value
+ * @return A EWM aggregation object
```
```suggestion
 * @return A EWMA aggregation object
```

bdice commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,330 @@
+}
+
+struct ewma_functor {
+  template <typename T, CUDF_ENABLE_IF(!std::is_floating_point<T>::value)>
```
```suggestion
  template <typename T, CUDF_ENABLE_IF(!std::is_floating_point_v<T>)>
```

bdice commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,330 @@
+    CUDF_FAIL("Unsupported type for EWMA.");
+  }
+
+  template <typename T, CUDF_ENABLE_IF(std::is_floating_point<T>::value)>
```
```suggestion
  template <typename T, CUDF_ENABLE_IF(std::is_floating_point_v<T>)>
```

bdice commented on cpp/src/reductions/scan/ewm.cu:
```diff
@@ -0,0 +1,330 @@
+#include <thrust/transform_scan.h>
+
+namespace cudf {
+namespace detail {
```
Can we wrap some parts of this in an anonymous namespace? Anything that is not declared in a header should be anonymous.

bdice commented on python/cudf/cudf/core/window/ewm.py:
```diff
@@ -0,0 +1,200 @@
+        Not Supported
+    adjust : bool, default True
+        Controls assumptions about the first value in the sequence.
+        https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ewm.html
+        for details.
```
Use a Sphinx reference instead of a raw link. Something like
```suggestion
        See :meth:`pandas.DataFrame.ewm` for details.
```

Please check that this renders nicely in the docs.

bdice commented on python/cudf/cudf/core/window/ewm.py:
```diff
@@ -0,0 +1,200 @@
+    2  1.615385
+    3  1.615385
+    4  3.670213
+
```
Be consistent with blank lines or no blank lines.
```suggestion
```

bdice commented on python/cudf/cudf/core/window/ewm.py:
```diff
@@ -0,0 +1,200 @@
+        if not is_numeric_dtype(sr.dtype):
+            raise TypeError("No numeric types to aggregate")
+
+        # libcudf ewm has special casing for nulls only
+        # and come what may with nans. It treats those nulls like
+        # pandas does nans in the same positions mathematically.
+        # as such we need to convert the nans to nulls before
+        # passing them in.
```
Can we make this comment clearer? What does "come what may" mean here, precisely?

bdice commented on python/cudf/cudf/tests/test_ewm.py:
We need tests for all the error modes (unsupported arguments, invalid values, etc.), and probably more input types, too.

bdice reviewed (approved):
This PR has stalled for a long time. I would like to approve and merge as-is, but please follow up with a new PR to address the comments in this review.

brandon-b-miller commented:
> This PR has stalled for a long time. I would like to approve and merge as-is, but please follow up with a new PR to address the comments in this review.

No problem @bdice . It shouldn't be too big of a jump from here to `ewmvar` and `ewmstd`, how about we take care of some of this cleanup at the same time?

bdice commented:
> No problem @bdice . It shouldn't be too big of a jump from here to `ewmvar` and `ewmstd`, how about we take care of some of this cleanup at the same time?

I'd prefer a cleanup before expanding the feature set, to avoid stalling in review. 😉

brandon-b-miller commented:
/merge
